staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt_test.go

```diff
@@ -0,0 +1,439 @@
+	// can't init nested struct in initializer
+	for i := range priorityLevelConfigurations {
+		priorityLevelConfigurations[i].Name = fmt.Sprintf("plc-%d", i)
+	}
+
+	serverConcurrencyLimit := groups
+	requestWaitLimit := 5 * time.Second
+	requests := 100
+
+	countnum := len(flowSchemas)
+	counts := []int64{}
+	for range flowSchemas {
+		counts = append(counts, int64(0))
+	}
+
+	delegate := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		time.Sleep(500 * time.Millisecond)
+		for i := 0; i < countnum; i++ {
+			if r.Header.Get("PRIORITY") == strconv.Itoa(i) {
+				atomic.AddInt64(&counts[i], int64(1))
+			}
+		}
+	})
+
+	server := createRequestManagementServer(
+		delegate, flowSchemas, priorityLevelConfigurations,
+		serverConcurrencyLimit, requestWaitLimit,
+	)
+	defer server.Close()
+
+	ws := []*Work{}
+	for i := 0; i < countnum; i++ {
+		header := make(http.Header)
+		header.Add("PRIORITY", strconv.Itoa(i))
+		req, _ := http.NewRequest("GET", server.URL, nil)
+		req.Header = header
+
+		ws = append(ws, &Work{
+			Request: req,
+			N:       requests,
+			C:       serverConcurrencyLimit,
+		})
+
+	}
+	for _, w := range ws {
+		w := w
+		// fmt.Printf("w %d info: %s\n", i, w.Request.Header.Get("PRIORITY"))
+		go func() { w.Run() }()
+	}
+
+	time.Sleep(1 * time.Second)
+	for i, count := range counts {
+		if count != 1 {
+			t.Errorf("Expected to dispatch 1 request for Group %d, found %v", i, count)
+		}
+	}
+}
+
+// TestRequestTimeout
+func TestRequestTimeout(t *testing.T) {
+	flowSchemas := []*rmtypesv1a1.FlowSchema{
+		&rmtypesv1a1.FlowSchema{
+			// metav1.TypeMeta
+			// metav1.ObjectMeta
+			Spec: rmtypesv1a1.FlowSchemaSpec{
+				PriorityLevelConfiguration: rmtypesv1a1.PriorityLevelConfigurationReference{
+					"plc-0",
+				},
+				// MatchingPrecedence: 1, // TODO(aaron-prindle) currently ignored
+				// DistinguisherMethod: *rmtypesv1a1.FlowDistinguisherMethodByUserType, // TODO(aaron-prindle) currently ignored
+				// Rules: []rmtypesv1a1.PolicyRuleWithSubjects{}, // TODO(aaron-prindle) currently ignored
+			},
+			// Status: rmtypesv1a1.FlowSchemaStatus{},
+		},
+	}
+
+	priorityLevelConfigurations := []*rmtypesv1a1.PriorityLevelConfiguration{
+		&rmtypesv1a1.PriorityLevelConfiguration{
+			Spec: rmtypesv1a1.PriorityLevelConfigurationSpec{
+				AssuredConcurrencyShares: 10,
+				Exempt:                   newFalse(),
+				GlobalDefault:            newFalse(),
+				HandSize:                 newHandSize(1),
+				QueueLengthLimit:         int32(65536),
+				Queues:                   int32(1),
+			},
+		},
+	}
+	for i := range priorityLevelConfigurations {
+		priorityLevelConfigurations[i].Name = fmt.Sprintf("plc-%d", i)
+	}
+
+	serverConcurrencyLimit := 1000
+	// TODO(aaron-prindle) currently ignored
+	requestWaitLimit := 1 * time.Nanosecond
+	requests := 5000
+
+	var count int64
+	delegate := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		atomic.AddInt64(&count, int64(1))
+		time.Sleep(1 * time.Second)
```

Comment:
We should not use real time in tests.  We should arrange our tests so that a fake clock can be used, and advanced as soon as the computation for each time is finished.
See https://github.com/MikeSpreitzer/kubernetes/blob/add-event-clock/staging/src/k8s.io/apimachinery/pkg/util/clock/event_clock.go for a clock that makes that easy.

Reply:
I have not completed this change but definitely agree and am in the process of making testing use a fake clock.